frontend/server/src/Controllers/Grader.php

```diff
@@ -25,19 +26,27 @@ public static function apiStatus(\OmegaUp\Request $r): array {
         ];
     }
     /**
-     * @psalm-return array{templateProperties: array{title: \OmegaUp\TranslationString, fullWidth?: bool, hideFooterAndHeader?: bool, payload: array<string, mixed>}, entrypoint: string, inContest?: bool, navbarSection?: string}
+     * @return array{templateProperties: array{payload: FullIDEPayload, title: \OmegaUp\TranslationString, fullWidth?: bool, hideFooterAndHeader?: bool}, entrypoint: string}
+     * @omegaup-request-param null|string $auth_token
     */
     public static function getGraderForTypeScript(
         \OmegaUp\Request $r
     ): array {
+        $preferredLanguage = \OmegaUp\DAO\Users::getPreferredLanguage(
+            \OmegaUp\Controllers\Session::getCurrentSession()['user']->user_id ?? null
+        );
```

Comment:
Okay, we have a more sophisticated way to get the preferred language for a user/identity:
```suggestion
        try {
            $r->ensureIdentity();
        } catch (\OmegaUp\Exceptions\UnauthorizedException $_) {
            // Do nothing. Not logged user can access here
            $r->identity = null;
        }

        $preferredLanguage = \OmegaUp\Controllers\Identity::getPreferredLanguage(
            $r->identity,
            $r
        );
```

In the first part, it ensures an identity exists, but it’s not required to have an open session. So, the `catch` part allows Not-logged users to access the ephemeral grader page.

As you can see, in the second part we have the function `Identity::getPreferredLanguage` that receives two parameters:

- `$request`: If the request contains the parameter `lang` (meaning the language is provided via URL) it becomes the preferred language.  
- `$identity`: This parameter is useful for getting the language that the user selected in their profile settings

Reply:
Yes I noticed that function but I think this function returns a spoken language not a programming language, right?

Reply:
Apologies. You're right. It's kind of difficult when we use the same term for different things. 

Maybe in the future we need to rename "language" to "programming language" (I think we already use it in some places).

Regarding the suggestion, you can obtain the `user_id` in a different way, avoiding access to the current session:

You can use the first part of my original suggestion in order to get the logged user:
```php
        $r->user = null;
        try {
            $r->ensureIdentity();
        } catch (\OmegaUp\Exceptions\UnauthorizedException $_) {
            // Do nothing. Not logged user can access here
            $r->user = null;
        }
```
And then, you can use `$r->user?->user_id` in the function argument:
```php
        $preferredLanguage = \OmegaUp\DAO\Users::getPreferredLanguage(
            $r->user?->user_id
        );
```

This way, you don't need to define the request param `$auth_token`, so it can be removed